### ISSUE TITLE
JSDoc type reference resolves incorrectly if checked before the constructor function it refers to

### ISSUE THREAD

**Issue by sandersn** [microsoft/TypeScript#34685]
```js
// @filename: mod1.js
var Workspace = {}
Workspace.Project = function () { }
Workspace.Project.prototype = {
  isServiceProject() { }
}
// @filename: test.js
/** @param {Workspace.Project} p */
function demo(p) {
  p.isServiceProject()
}
```

Expected: no error, and `p.isServiceProject(): void`.

Actual: no error if you compile `tsc mod1.js test.js`, but an error with `tsc test.js mod1.js`:

Property 'isServiceProject' does not exist on type 'typeof Project'.

From chrome-devtools-frontend, failed in https://github.com/microsoft/TypeScript/pull/34603

**Comment by sandersn** [comment]
Also broken by #34515, but for a different reason.
